gunfolds/utils/clingo.py

```diff
@@ -41,26 +41,45 @@ def run_clingo(command,
         trendy: Use defaults geared towards industrial problems
     :type configuration: string
 
-    :param cpath: clingo path
-    :type cpath: string
 
     :param pnum: number of parallel threads to run clingo on
     :type pnum: integer
 
+    :param optim: a list containing configuration for optimization algorithm and optionally a bound  [<arg>, <bound>]
+        <arg>: <mode {opt|enum|optN|ignore}>[,<bound>...]
+        opt   : Find optimal model
+        enum  : Find models with costs <= <bound>
+        optN  : Find optimum, then enumerate optimal models
+        ignore: Ignore optimize statements
+        <bound> : Set initial bound for objective function(s)
+    :type optim: list
+
     :returns: results of equivalent class
     :rtype: dictionary
     """
+    if optim is None:
+        optim = ['optN']
     if pnum is None:
         pnum = PNUM
-    ctrl = clngo.Control(["--warn=no-atom-undefined","--configuration=", configuration, "-t", str(int(pnum)) + ",split", "-n", str(capsize)])
+    if len(optim) == 1:
+        optim_option = optim[0]
+    elif len(optim) == 2:
+        optim_option = f"{optim[0]}, {optim[1]}"
+    else:
+        raise ValueError("The 'optim' list must have either one or two elements.")
+
+    clingo_control = ["--warn=no-atom-undefined", "--configuration=", configuration, "-t", str(int(pnum)) + ",split"]
+    if capsize is not None:
+        clingo_control += ["-n", str(capsize)]
+    ctrl = clngo.Control(clingo_control)
     if not exact:
-        ctrl.configuration.solve.opt_mode = "opt"
+        ctrl.configuration.solve.opt_mode = optim_option
     ctrl.add("base", [], command.decode())
     ctrl.ground([("base", [])])
     models = []
     with ctrl.solve(yield_=True, async_=True) as handle:
         for model in handle:
-            models.append([str(atom) for atom in model.symbols(shown=True)])
+            models.append(([str(atom) for atom in model.symbols(shown=True)], model.cost))
     cost = ctrl.statistics["summary"]["costs"]
     num_opt = ctrl.statistics["summary"]["models"]["optimal"]
     if not exact:
```

Comment:
is there a way to reduce the number of if then else? This syntax sugar is hurting my eyes :) I understand this works, but can we have sensible defaults in at least some of those cases? In this case the code will not be polluted

Reply:
we can remove this if statement:
  ```
  if pnum is None:
        pnum = PNUM
```
        but this if statement I think is required :

```    if capsize is not None:
        clingo_control += ["-n", str(capsize)]```        
        Clingo behaves different when we don't give it this parameter at all Vs. when we set `-n 0`